### ISSUE TITLE
FSharp.Core version 6.0.5 has multiple content hashes.

### ISSUE THREAD

**Issue by JustinWilkinson** [dotnet/fsharp#13678]
Please provide a succinct description of the issue.
Version `6.0.5` of the `FSharp.Core` library seems to have two different contents depending on where it is delivered from.
The version bundled in the SDK (found under `C:\Program Files\dotnet\sdk\6.0.400\FSharp\library-packs` stems from [this commit](https://github.com/dotnet/fsharp/commit/1b94f89d4d1f41f20f9be73c76f4b229d4e49078).
![image](https://user-images.githubusercontent.com/36571008/184317932-abce9976-2f57-43fa-ac3c-53f3a31b1d21.png)

The version published on [NuGet.org](https://www.nuget.org/packages/FSharp.Core/6.0.5) stems from [this commit](https://github.com/dotnet/fsharp/commit/211fd760c2315d73cfa0d375fe64abb346bc1096).
![image](https://user-images.githubusercontent.com/36571008/184318011-fbdded0c-0443-43bf-81a3-c21d2f576db4.png)

In our builds, we make use of `packages.lock.json` and `global.json` files to ensure repeatable builds across different machines.

We make use of the below in a `Directory.Build.props` file to ensure hashes are the same, and this has worked historically up until the `6.0.400` SDK release.
```
<RestorePackagesWithLockFile>true</RestorePackagesWithLockFile>
<DisableImplicitNuGetFallbackFolder>true</DisableImplicitNuGetFallbackFolder>
```

The version on `NuGet.org` predates the version in the SDK, so the content hash on local builds seems to update but our pipelines are failing with the error:

> error NU1403: Package content hash validation failed for FSharp.Core.6.0.5. The package is different than the last restore.

Given the hashes (and the content) are indeed different, this is expected, what is unexpected is the delivery of two distinct packages with the same version number.

**Repro steps**

1. Create a project that references `FSharp.Core` version 6.0.5, enable lock files and restore.
2. Install the `6.0.400` SDK.
3. Restore packages again.

**Expected behavior**
The content of the `FSharp.Core` package (version `6.0.5`) delivered by the `6.0.400` SDK should match that on `NuGet.org`.

**Actual behavior**

The content of the `FSharp.Core` package (version `6.0.5`) delivered by the `6.0.400` SDK is different to that which is delivered on `NuGet.org`.

**Known workarounds**

We have downloaded the package from `NuGet.org` and assigned it version `6.0.5.1` in our own feed to force it to pick up a new version.

* Operating system - Windows
* .NET Runtime kind - .NET 6.
* Editing Tools - Visual Studio version 17.3

**Comment by vzarytovskii** [comment]
@KevinRansom it seems that it is because we didn't bump a version

**Comment by cmeeren** [comment]
All our builds have started failing with this error now. Are there any easier/better workarounds than what @JustinWilkinson described? We'd have to do this for hundreds of projects. (Turning off lock files is not an option at the moment.)

**Comment by cmeeren** [comment]
I tried pinning the .NET SDK version to 6.0.303 using global.json, but I get the same result. What gives?

**Comment by hcoona** [comment]
I also meet this issue

**Comment by cmeeren** [comment]
Reverting to 6.0.302 worked. 6.0.303 did not work.

**Comment by vzarytovskii** [comment]
I think it is an issue with that we didn't bump the version when we were getting the localization changes in, and it went to SDK but not to the NuGet.

**Comment by vzarytovskii** [comment]
Is removing FSharp.Core (6.0.5) from local nuget cache and having `<DisableImplicitNuGetFallbackFolder>true</DisableImplicitNuGetFallbackFolder>` helping the build.

**Comment by bddckr** [comment]
> Is removing FSharp.Core (6.0.5) from local nuget cache and having `<DisableImplicitLibraryPacksFolder>true</DisableImplicitLibraryPacksFolder>` helping the build?

It does indeed help. For my team, this reliably pulls down the version found on NuGet.org.

So the steps are
1. Ensure your project(s) use `<DisableImplicitLibraryPacksFolder>true</DisableImplicitLibraryPacksFolder>`.
1. Delete the locally cached copy of 6.0.5 out of your `.nuget` folder, i.e. delete the folder `C:\Users\<your windows user>\.nuget\packages\fsharp.core\6.0.5` on Windows.
1. Restore packages and update your lock files. This will write the other `contentHash` into the lock files.
1. Make those new lock files available to the rest of the team.
1. Ask every team member to wipe their cached copy followed by a restore as well.
1. If your CI caches restored packages, make sure to wipe the cached copy there as well.

**Comment by cmeeren** [comment]
A couple of questions:

1. What exactly is `DisableImplicitLibraryPacksFolder`? Web searches turn up nothing useful about this. I'm hesitant to use it (especially since I am now using the 6.0.302 SDK as a simple and reliable workaround).
2. In the steps posted by @bddckr above, would it cause problems if some projects/solutions on your machine did not use `DisableImplicitLibraryPacksFolder`? Would they then try to place the package from the SDK in the NuGet cache (and win if you opened them first)?

**Comment by bddckr** [comment]
Apologies, I copied the wrong property and confused myself.

My team is using `DisableImplicitNuGetFallbackFolder` like the original poster said. This works fine - as long as every team member ensures to delete the cached version that has the "wrong" hash. We do _NOT_ use `DisableImplicitLibraryPacksFolder`.

I've edited my earlier message to fix that.

**Comment by bddckr** [comment]
Both properties are seemingly very closely related for the same thing - the extra search paths pointing to files not sourced from NuGet:
https://github.com/dotnet/sdk/blob/a30e465a2e2ea4e2550f319a2dc088daaafe5649/src/Tasks/Microsoft.NET.Build.Tasks/targets/Microsoft.NET.NuGetOfflineCache.targets#L20-L29

I'm thinking about enabling both moving forward. I believe `DisableImplicitLibraryPacksFolder` should be set to `true` for the same reason the recommendation is to enable `DisableImplicitNuGetFallbackFolder`:
https://github.com/dotnet/reproducible-builds/blob/d6b3923a0908249a8d6416a1e2d02c51e53b6cce/Documentation/Reproducible-MSBuild/Techniques/DisableImplicitNuGetFallbackFolder.md

**Comment by cmeeren** [comment]
> My team is using `DisableImplicitNuGetFallbackFolder` like the original poster said. This works fine - as long as every team member ensures to delete the cached version that has the "wrong" hash. We do _NOT_ use `DisableImplicitLibraryPacksFolder`.

Not sure what you mean by "works fine". I have long used `DisableImplicitNuGetFallbackFolder` (and not `DisableImplicitLibraryPacksFolder`), deleting the NuGet cache did not change the content hash in the lock files after `dotnet restore --force-evaluate`.

> I'm thinking about enabling both moving forward. I believe `DisableImplicitLibraryPacksFolder` should be set to `true` for the same reason the recommendation is to enable `DisableImplicitNuGetFallbackFolder`:

Makes sense. Would be great to have someone else confirm this.

**Comment by bddckr** [comment]
> Not sure what you mean by "works fine". I have long used `DisableImplicitNuGetFallbackFolder` (and not `DisableImplicitLibraryPacksFolder`), deleting the NuGet cache did not change the content hash in the lock files after `dotnet restore --force-evaluate`. Everything builds fine locally (either way), but the CI build fails.

Hmmm. I just set up a fresh Windows VM (for something unrelated) and tried it there as well. You're right. It keeps using the library-packs folder. I must have gotten my results wrong when I was testing `DisableImplicitLibraryPacksFolder` earlier as well.

Finally, to confirm (sorry for the back and forth): If I'm setting both `DisableImplicitNuGetFallbackFolder` and `DisableImplicitLibraryPacksFolder` to `true`, the original content hash is used - the hash I got from before the `6.0.400` SDK was installed. As such, the lock file stays the same, and I don't have to ask my team members to wipe their cached copy. I simply need to ensure they are on the latest project state - using `<DisableImplicitLibraryPacksFolder>true</DisableImplicitLibraryPacksFolder>`.

tl;dr: vzarytovskii's suggestion above works.

**Comment by cmeeren** [comment]
Happy to hear it.

Would be great to get confirmation from at least one other knowledgeable person that using `DisableImplicitLibraryPacksFolder` is a general good idea for the same reasons as `DisableImplicitNuGetFallbackFolder`.

**Comment by vzarytovskii** [comment]
> Happy to hear it.
> 
> Would be great to get confirmation from at least one other knowledgeable person that using `DisableImplicitLibraryPacksFolder` is a general good idea for the same reasons as `DisableImplicitNuGetFallbackFolder`.

It's fine to use it. It was made for scenarios like this - when there's need to turn off implicit packs import.

**Comment by cmeeren** [comment]
Thanks! Should it then perhaps be added alongside https://github.com/dotnet/reproducible-builds/blob/main/Documentation/Reproducible-MSBuild/Techniques/DisableImplicitNuGetFallbackFolder.md?

**Comment by baronfel** [comment]
I'm a maintainer on that project and would gladly accept a contribution outlining how and why to disable this property. Happy to help wordsmith it too!

**Comment by dsyme** [comment]
@vzarytovskii What's the status, is it point in time and can be closed, or do we need to fix it? Thanks

**Comment by vzarytovskii** [comment]
Yeah, this should be closed I think, nothing we can do for this particular version. Should be fixed for future ones (I will also have a CI gate in place to check it won't happen again).

**Comment by vzarytovskii** [comment]
> @vzarytovskii - not sure if the CI gate's worked.
> 
> We'd swapped to using `DisableImplicitLibraryPacksFolder` in most cases but a couple were missed and we've detected this issue again for version `6.0.6` distributed in the 6.0.402 SDK has this same issue, in that the commit hash is different to the one on NuGet
> @dsyme 
> ![image](https://user-images.githubusercontent.com/36571008/195369605-b78fe143-4008-4016-a7d8-166a244fc315.png)
> ![image](https://user-images.githubusercontent.com/36571008/195369679-6a173e8f-32a7-442b-8f13-ffced9568f20.png)
> 

CI gate is not in place yet.

@KevinRansom we again have different hashes for the package in nuget and SDK

**Comment by vzarytovskii** [comment]
Sigh..., it seems it is, in fact, caused by the differences in how insertions work for VS and SDK, and the package was published from VS pipeline, which has one less insertion.

**Comment by vzarytovskii** [comment]
We'll probably need to change the flow - how and from where do we publish nugets.

**Comment by KevinRansom** [comment]
sdk publishes them, and given that, I have no idea how they could mess that up.

**Comment by KevinRansom** [comment]
Okay, I believe I know what this is.  It highlighted a bug in the sdk publish process that caused them to not update the package in the upload cache.  I believe this release is the one they discovered it with and it will be fixed in the next release.  I will follow up with @mmitche later today and report back if necessary.

**Comment by petteripoyhtari** [comment]
We have now currently this same issue with the 7.0.0 package and we have trouble to pass our Azure ci/cd pipeline builds through. Some machines will get the same hash and other won't.

Could you change the package version number when you change the package contents / code, please.
![fsharp-7-hash-fail](https://user-images.githubusercontent.com/81987998/212839873-ec0a11a7-daf9-4df2-ad03-376826f3d010.png)
